pkg/app/carbonapi/http_handlers.go

```diff
@@ -940,38 +938,128 @@ func (app *App) findHandler(w http.ResponseWriter, r *http.Request, lg *zap.Logg
+	Trace(lg, "received request")
+
+	logLevel := zap.InfoLevel
+	defer func() {
+		app.deferredAccessLogging(lg, r, &toLog, t0, logLevel)
+	}()
+
+	if query == "" {
+		writeError(uuid, r, w, http.StatusBadRequest, "missing parameter `query`", "", &toLog)
+		return
+	}
+	metrics, fromCache, err := app.resolveGlobs(ctx, query, useCache, &toLog, lg)
+	toLog.FromCache = fromCache
+	if err == nil {
+		toLog.TotalMetricCount = int64(len(metrics.Matches))
+	} else {
+		var notFound dataTypes.ErrNotFound
+
+		switch {
+		case errors.As(err, &notFound):
+			// we can generate 404 for expand, it's graphite-web-1.0 only
+			Trace(lg, "not found")
+			writeError(uuid, r, w, http.StatusNotFound, err.Error(), "", &toLog)
+			logLevel = zapcore.ErrorLevel
 			return
-		}
-		if _, writeErr := w.Write([]byte{'('}); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
+		case errors.Is(err, context.DeadlineExceeded):
+			writeError(uuid, r, w, http.StatusUnprocessableEntity, "context deadline exceeded", "", &toLog)
+			logLevel = zapcore.ErrorLevel
 			return
-		}
-		if _, writeErr := w.Write(blob); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
+		default:
+			writeError(uuid, r, w, http.StatusUnprocessableEntity, err.Error(), "", &toLog)
+			logLevel = zapcore.ErrorLevel
 			return
 		}
-		if _, writeErr := w.Write([]byte{')'}); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
-			return
+	}
+
+	if ctx.Err() != nil {
+		app.ms.RequestCancel.WithLabelValues("expand", ctx.Err().Error()).Inc()
+	}
+
+	blob, err := expandEncoder(metrics, leavesOnly, groupByExpr)
+	if err != nil {
+		writeError(uuid, r, w, http.StatusInternalServerError, err.Error(), "", &toLog)
+		logLevel = zapcore.ErrorLevel
+		return
+	}
+	if writeResponse(ctx, w, blob, jsonFormat, jsonp) != nil {
+		toLog.HttpCode = 499
+		logLevel = zapcore.ErrorLevel
+		return
+	}
+}
+
+func expandEncoder(globs dataTypes.Matches, leavesOnly bool, groupByExpr bool) ([]byte, error) {
+	var b bytes.Buffer
+	groups := make(map[string][]string)
+	seen := make(map[string]struct{})
+	nodeCount := len(strings.Split(globs.Name, "."))
+	names := make([]string, 0, len(globs.Matches))
+	for _, g := range globs.Matches {
+		if leavesOnly && !g.IsLeaf {
+			continue
 		}
-	} else {
-		w.Header().Set("Content-Type", contentType)
-		if _, writeErr := w.Write(blob); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
-			return
+		name := g.Path
+		nodes := strings.SplitN(name, ".", nodeCount+1)
+		if len(nodes) > nodeCount {
+			name = strings.Join(nodes[:nodeCount], ".")
+		} else {
+			name = strings.Join(nodes, ".")
+		}
+		if _, ok := seen[name]; ok {
+			continue
 		}
+		seen[name] = struct{}{}
+		names = append(names, name)
 	}
-
-	toLog.HttpCode = http.StatusOK
+	sort.Strings(names)
+	groups[globs.Name] = names
+	data := map[string]interface{}{
```

Comment:
I may be wrong, but I don't understand why this should be a `map[string]interface{}`. Looks like we can have an _if_ that does a separate encoding for both values of `groupByExpr`. We can then use specific data structures in both branches of _if_. I'm not sure if and how much it will improve performance (but I'd guess there will be a performance gain, because this way we will likely avoid an additional reflection during serialization), but it will certainly make the code much more clear.